crates/fuel-core/src/service/adapters/producer.rs

```diff
@@ -123,31 +123,53 @@ impl fuel_core_producer::ports::DryRunner for ExecutorAdapter {
 
 #[async_trait::async_trait]
 impl fuel_core_producer::ports::Relayer for MaybeRelayerAdapter {
-    async fn wait_for_at_least(
+    async fn get_latest_da_blocks_with_costs(
         &self,
-        height: &primitives::DaBlockHeight,
-    ) -> anyhow::Result<primitives::DaBlockHeight> {
+        starting_from: &DaBlockHeight,
+    ) -> anyhow::Result<Vec<(DaBlockHeight, u64)>> {
         #[cfg(feature = "relayer")]
         {
             if let Some(sync) = self.relayer_synced.as_ref() {
-                sync.await_at_least_synced(height).await?;
-                sync.get_finalized_da_height()
+                sync.await_at_least_synced(starting_from).await?;
+                let highest = sync.get_finalized_da_height()?;
+                (starting_from.0..=highest.0)
+                    .map(|height| get_gas_cost_for_height(height, sync))
+                    .collect()
             } else {
-                Ok(0u64.into())
+                Ok(Vec::new())
             }
         }
         #[cfg(not(feature = "relayer"))]
         {
             anyhow::ensure!(
-                **height == 0,
+                **starting_from == 0,
                 "Cannot have a da height above zero without a relayer"
             );
             // If the relayer is not enabled, then all blocks are zero.
-            Ok(0u64.into())
+            Ok(Vec::new())
         }
     }
 }
 
+#[cfg(feature = "relayer")]
+fn get_gas_cost_for_height(
+    height: u64,
+    sync: &fuel_core_relayer::SharedState<
+        Database<crate::database::database_description::relayer::Relayer>,
+    >,
+) -> anyhow::Result<(DaBlockHeight, u64)> {
+    let da_height = DaBlockHeight(height);
+    let cost = sync
+        .database()
+        .storage::<fuel_core_relayer::storage::EventsHistory>()
+        .get(&da_height)?
+        .map(|cow| cow.into_owned())
```

Comment:
You don't need to call `into_owned` here. It is the place where we want to use references if possible.